kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -109,30 +113,37 @@ internal open class ObjCExportNameTranslatorImpl(
     override fun getCategoryName(file: KtFile): String =
             helper.translateFileName(file)
 
-    override fun getClassOrProtocolName(
-            ktClassOrObject: KtClassOrObject
-    ): ObjCExportNamer.ClassOrProtocolName = helper.swiftClassNameToObjC(
-            getClassOrProtocolSwiftName(ktClassOrObject)
-    )
+    override fun getClassOrProtocolName(ktClassOrObject: KtClassOrObject): ObjCExportNamer.ClassOrProtocolName =
+            ObjCExportNamer.ClassOrProtocolName(
+                    swiftName = getClassOrProtocolName(ktClassOrObject, true),
+                    objCName = buildString {
+                        append(configuration.topLevelNamePrefix)
```

Comment:
Correct. I think you will always want the top-level prefix since that is the only way to prevent conflicts with other libraries that have classes with the same names.

Reply:
Ok.
I guess this is one more case where it is just the simplest solution. As I [said](https://github.com/JetBrains/kotlin/pull/4815#discussion_r876680000), this works for me. Also, this affects only Objective-C.

Reply:
This would indeed solve the problem.
What should this parameter do for functions?

Reply:
> I could imagine situations (gradually moving module to Kotlin and wanting to keep exact names the same in Objective-C) where you will want to create a precise Objective-C name without anything added to what is in the annotation.

Good point. How about making this `prefixObjCName` a compiler option?
Not sure about the name but something like: `-exact-objc-name`. 
If it's set the `ObjCName` values are interpreted as exact values, otherwise we still use the logic to add top level names, parent containers, etc.

Reply:
Actually a compiler option wouldn't work as it can't be controlled by libraries or included modules.
So that leaves us with 2 options:
- add a parameter to the annotation
- add another annotation to make `CName` values exact

Neither are perfect IMO.
Adding a parameter to the annotation means it's also available when e.g. the annotation is used on a function parameter. 
In thay case the behavior is always the same since those names are already exact values.
That could be solved by another annotation which is only available for classes. But that means another annotation...

What do you think?

Reply:
I think my personal preference is to have only one annotation with a parameter and then document where it has meaning via the name of the parameter and doc comments on the annotation.

I like the term "ensure exact objective-c name". Maybe the annotation parameter could be called `exactObjCName`. If the default for that is `false`, it seems clearer that explicitly setting that to `true` will not have an effect on functions where the name is already exact. Then in doc comments we can document when exactly this has effect and as importantly, where it doesn't?

Reply:
Agreed! It wouldn't make sense to only make the ObjC name exact, but not the Swift name, right?
How about?:
```kotlin
@ObjCName(name = "ObjCName, swiftName = "SwiftName", exact = true)
```

Reply:
> I like the term "ensure exact objective-c name". Maybe the annotation parameter could be called `exactObjCName`. If the default for that is `false`, it seems clearer that explicitly setting that to `true` will not have an effect on functions where the name is already exact.

This might be a bit confusing -- one can think that "exact" here means that the compiler doesn't mangle this name in the presence of clashes.

Reply:
> > I like the term "ensure exact objective-c name". Maybe the annotation parameter could be called `exactObjCName`. If the default for that is `false`, it seems clearer that explicitly setting that to `true` will not have an effect on functions where the name is already exact.
> 
> This might be a bit confusing -- one can think that "exact" here means that the compiler doesn't mangle this name in the presence of clashes.

I wonder if that shouldn't be the actual meaning of it? If you specify that you want an exact objective-c name, any other name will probably break what you are trying to do. Therefore, a conflict with an exact objective-c name should probably be a compilation error and not rename?

Reply:
Can't we use the name that is best for the final design? The annotation will be experimental, so it's expected to not be 100% complete/final, right?

Reply:
> The annotation will be experimental, so it's expected to not be 100% complete/final, right?

Yes.

> Well it describes the correct behaviour.

Sorry, I'm a bit lost. What should happen if a user specifies `@ObjCName(..., exact = false)`, and the compiler has to mangle the name?
Is there a way to prohibit mangling but keep the Objective-C class name prefix? Except for adding the prefix to the name explicitly.

Reply:
Ok.
Let's add this new annotation parameter that will only control whether the top-level prefix is added to the name, but make the frontend checker ensure that this parameter is not used on anything but classes?

Reply:
> Let's add this new annotation parameter that will only control whether the top-level prefix is added to the name, but make the frontend checker ensure that this parameter is not used on anything but classes?

Sounds good 👍🏻